design/resource-dependencies.md

```diff
@@ -0,0 +1,118 @@
+# Resource dependency relationships
+
+### Inventory of resources that could have dependency relationships
+All Rook-Ceph CRDs:
+- CephCluster
+- CephBlockPool
+- CephFilesystem
+- CephObjectStore
+- CephNFS
+- CephClient
+- CephRBDMirror
+- CephFilesystemMirror
+- CephObjectRealm
+- CephObjectZoneGroup
+- CephObjectZone
+
+Resources for which Rook-Ceph acts as a driver/provisioner:
+- ObjectBucketClaim (lib-bucket-provisioner)
+- ObjectBucket (COSI)
+
+Kubernetes resources which can depend on Rook:
+- StorageClass
+- PersistentVolume
+
+## Proposed dependency relationships
+
+#### `CephCluster`
+Primary dependents which can create/delete pools:
+- CephBlockPools in the same namespace
+- CephFilesystems in the same namespace
+- CephObjectStores in the same namespace
+
+Dependents which can consume arbitrary pools including user-created pools:
+- CephNFSes in the same namespace
+- CephClients in the same namespace
+
+#### `CephBlockPool`
+Dependents which can consume pools:
+- CephNFS in the same namespace that have `spec.pool == <a provider pool>`
+- CephClient in the same namespace that has any `spec.caps` value with the string `pool=<a provider pool>`
+
+Dependents via CSI:
+- StorageClasses that have:
+  - `provisioner == <operator-namespace>.rbd.csi.ceph.com` AND
+  - `parameters.clusterID == <ceph-cluster-namespace>` AND
+  - `parameters.pool` OR `parameters.dataPool` references a `CephBlockPool` pool
+- PersistentVolumes that have:
+  - `spec.CSI.Driver` == `<operator-namespace>.rbd.csi.ceph.com` AND
+  - `spec.CSI.VolumeAttributes["clusterID"] == <ceph-cluster-namespace>` AND
+  - `spec.CSI.VolumeAttributes["pool"]` OR `spec.CSI.VolumeAttributes["journalPool"]` references a `CephBlockPool` pool
+  - OR
+  - `spec.storageClassName == <name of StorageClass which references the CephBlockPool>`
+
+Dependents via FlexVolume:
+- StorageClasses that have:
+- PersistentVolumes that have:
+
+TODO: decide if we should treat PVs OR StorageClasses as dependents. IMO, in following with the rest
```

Comment:
Traversing PVs and storage classes feels out of scope to me. There could be thousands of PVs that we don't even own. Instead of querying PVs and storage classes, this is where it seems correct to query Ceph if there are related resources.

Reply:
If we really needed to guarantee correctness, sure, traversing all those could make sense. That level of completeness just shouldn't be necessary for cluster cleanup IMO.

Reply:
Currently, we do list PVs when deleting the CephCluster but not StorageClasses.

Based on what we've been talking about and the scheme proposed here, I think I agree with you about PV traversal not being needed for cluster cleanup. Relatively speaking, I think it is _more_ important to check PVs when deleting CephBlockPools and CephFilesystems than when deleting CephClusters since those actually provision the type of storage which the PV will get and since the CephCluster has block/file as dependents. However, it is a good point that it could be a lot of work to traverse StorageClasses and PVs in large clusters.

I'm not sure querying Ceph would be able to tell us if there are PVs. We could check to see if there are images in a block pool or modifications to a filesystem, but that won't tell us if there are existing claims on block/file storage that haven't yet written data.

I don't think we **have** to search StorageClasses or PersistentVolumes to see if any reference storage when we are doing deletes. And we can definitely do this in a separate PR. I think it's worth discussing whether there is user benefit for Rook doing that traversal on deletion and whether it is appropriate for Rook to query those resources. 

For this doc, I would like to capture what is needed in order to inspect PV/SC dependencies so that we can see the complexity and weigh whether we want to include these as dependents in our design plans.